[Community Pipeline] IPAdapter FaceID

fabiorigano opened the pull request:
# What does this PR do?
Add support for IPAdapter FaceID
Fixes #6243 

## Who can review?

@patrickvonplaten, @sayakpaul, @yiyixuxu

fabiorigano commented:
```
import cv2
from insightface.app import FaceAnalysis
import numpy as np
from PIL import Image
import torch
from diffusers import StableDiffusionPipeline, DDIMScheduler
from diffusers.utils import load_image

def image_grid(imgs, rows, cols):
    assert len(imgs) == rows*cols

    w, h = imgs[0].size
    grid = Image.new('RGB', size=(cols*w, rows*h))
    grid_w, grid_h = grid.size

    for i, img in enumerate(imgs):
        grid.paste(img, box=(i%cols*w, i//cols*h))
    return grid

noise_scheduler = DDIMScheduler(
    num_train_timesteps=1000,
    beta_start=0.00085,
    beta_end=0.012,
    beta_schedule="scaled_linear",
    clip_sample=False,
    set_alpha_to_one=False,
    steps_offset=1
)

pipeline = StableDiffusionPipeline.from_pretrained(
    "SG161222/Realistic_Vision_V4.0_noVAE",
    torch_dtype=torch.float16,
    scheduler=noise_scheduler,
    feature_extractor=None,
    safety_checker=None
).to("cuda")

generator = torch.Generator(device="cpu").manual_seed(42)
num_images=4
image = load_image("https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/ai_face2.png")

app = FaceAnalysis(name="buffalo_l", providers=['CUDAExecutionProvider', 'CPUExecutionProvider'])
app.prepare(ctx_id=0, det_size=(640, 640))

image = cv2.cvtColor(np.asarray(image), cv2.COLOR_BGR2RGB)
faces = app.get(image)
image = torch.from_numpy(faces[0].normed_embedding).unsqueeze(0)

pipeline.load_ip_adapter("h94/IP-Adapter-FaceID", weight_name="ip-adapter-faceid_sd15.bin")
pipeline.set_ip_adapter_scale(0.7)
images = pipeline(
  prompt="A photo of a girl wearing a black dress, holding red roses in hand, upper body, behind is the Eiffel Tower",
  image_embeds=image,
  negative_prompt="monochrome, lowres, bad anatomy, worst quality, low quality", 
  num_inference_steps=20, num_images_per_prompt=num_images, width=512, height=704, 
  generator=generator,
  #output_type= "np"
).images
image_grid(images, 1, 4)
```

##Output
![image](https://github.com/huggingface/diffusers/assets/57982783/cc5ec259-b349-4fa1-b583-31bb43f3c97f)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_6276). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

a-r-r-o-w commented:
Thank you so much for adding this @fabiorigano! Was trying to work on this as well but great to see the quick addition. Sharing some more results from the PR with EpicRealism-v4.

<details>
<summary>Results</summary>
<table>
<tr>
  <td colspan="3">Original Image</td>
</tr>
<tr>
  <td colspan="3"><img src="https://github.com/huggingface/diffusers/assets/72266394/3186693a-17cc-4441-9963-19966a37d0cb"></td>
</tr>
<tr>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/ff016b6c-e423-4395-9930-58e786d9e46e"></td>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/349be0a5-3643-491a-80a6-25f212c90875"></td>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/f0a15f0a-ba2e-4714-af88-a0055147a248"></td>
</tr>
<tr>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/27c0ea03-8a94-4d70-af33-48f2318c5ed2"></td>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/1e73c703-6738-4fcb-824f-1a1afe410afd"></td>
  <td><img src="https://github.com/huggingface/diffusers/assets/72266394/ce21f7c6-10b0-4ad8-92f1-44320bf0d724"></td>
</tr>
</table>
</details>

sayakpaul commented:
@fabiorigano thanks so much for this. 

Let's try to get the CI green.

sayakpaul commented on src/diffusers/loaders/ip_adapter.py:
```diff
@@ -46,8 +48,8 @@ class IPAdapterMixin:
         pretrained_model_name_or_path_or_dict: Union[str, Dict[str, torch.Tensor]],
-        subfolder: str,
         weight_name: str,
+        subfolder: Optional[str] = None,
```
We can pass this to the `kwargs` right? I don't think there's a need to expose this specifically. WDYT?

fabiorigano replied:
> We can pass this to the `kwargs` right? I don't think there's a need to expose this specifically. WDYT?

Yes, we can pass it as keyword argument. The other IP Adapter models need the image encoder, stored in a subfolder of H94/IP-Adapter, while FaceID model doesn't require an image encoder, so I made it Optional. I will remove it in the next update

sayakpaul commented on src/diffusers/loaders/unet.py:
```diff
@@ -684,13 +684,20 @@ def _convert_ip_adapter_image_proj_to_diffusers(self, state_dict):
                 updated_state_dict[diffusers_name] = value
 
-        elif "proj.3.weight" in state_dict:
+        elif "proj.0.weight" in state_dict:
```
Is using this key a better option? Can we use a more resilient condition here to avoid side-effects?

fabiorigano replied:
I changed to `proj.0.weight` because both IPAdapter Full and FaceID state dicts have it, while the IPAdapter Full `proj.3.weight` key is named `norm.weight` in the FaceID model

sayakpaul commented:
Would be also nice to have some documentation about it here:
https://huggingface.co/docs/diffusers/main/en/using-diffusers/loading_adapters#ip-adapter

WDYT?

sayakpaul commented on src/diffusers/loaders/unet.py:
```diff
@@ -774,20 +791,47 @@ def _load_ip_adapter_weights(self, state_dict):
+                        hidden_size=hidden_size,
+                        cross_attention_dim=cross_attention_dim,
+                        scale=1.0,
+                        rank=128,
```
Do we need to make `rank` an argument?

fabiorigano replied:
Not for now, but perhaps new IP Adapter will be released in the future, using different LoRA ranks. Do you think it is better to remove it for now?

fabiorigano commented:
> Would be also nice to have some documentation about it here: https://huggingface.co/docs/diffusers/main/en/using-diffusers/loading_adapters#ip-adapter
> 
> WDYT?

Sure, I can add the snippet on top of this PR

patrickvonplaten commented:
@yiyixuxu could you take a look here?

patrickvonplaten commented on src/diffusers/loaders/ip_adapter.py:
```diff
@@ -46,7 +48,6 @@ class IPAdapterMixin:
     def load_ip_adapter(
         self,
         pretrained_model_name_or_path_or_dict: Union[str, Dict[str, torch.Tensor]],
-        subfolder: str,
```
This is a breaking change, can we leave as it was here?

fabiorigano replied:
done as you suggested!

patrickvonplaten commented on src/diffusers/loaders/ip_adapter.py:
```diff
@@ -135,14 +137,15 @@ def load_ip_adapter(
-                self.image_encoder = image_encoder
-            else:
-                raise ValueError("`image_encoder` cannot be None when using IP Adapters.")
+                try:
+                    logger.info(f"loading image_encoder from {pretrained_model_name_or_path_or_dict}")
+                    image_encoder = CLIPVisionModelWithProjection.from_pretrained(
+                        pretrained_model_name_or_path_or_dict,
+                        subfolder=os.path.join(subfolder, "image_encoder"),
+                    ).to(self.device, dtype=self.dtype)
+                    self.image_encoder = image_encoder
+                except TypeError:
+                    print("IPAdapter: `subfolder` not found, `image_encoder` is None, use image_embeds.")
```
```suggestion
                try:
                    logger.info(f"loading image_encoder from {pretrained_model_name_or_path_or_dict}")
                    image_encoder = CLIPVisionModelWithProjection.from_pretrained(
                        pretrained_model_name_or_path_or_dict,
                        subfolder=os.path.join(subfolder, "image_encoder"),
                    ).to(self.device, dtype=self.dtype)
                    self.image_encoder = image_encoder
                except TypeError:
                    print("IPAdapter: `subfolder` not found, `image_encoder` is None, use image_embeds.")
```

Let's try to not use `try...except` here please

patrickvonplaten commented:
@yiyixuxu can you take a look here?

a-r-r-o-w commented:
Btw, [FaceID v2 and SDXL](https://huggingface.co/h94/IP-Adapter-FaceID) versions were made public last week. Maybe we could look into supporting and testing both in this PR too?

sayakpaul commented:
One at a time please.

fabiorigano commented:
I just fixed LoRA weights loading, because some of them were missing.
With this update, output images are more aligned with input image, thanks to both LoRA and IP Adapter.

![_p0](https://github.com/huggingface/diffusers/assets/57982783/a92c1b9d-8f00-4cd3-bb03-158d71985bd9)
![_p1](https://github.com/huggingface/diffusers/assets/57982783/ecd8a7b9-7d0a-4b50-9cdf-af672adab45e)


I updated set_default_attn_processor in order to restore processors with torch sdpa.

yiyixuxu commented:
@patrickvonplaten 
This is still an experimental feature - does creating a community pipeline for this make more sense?

yiyixuxu reviewed (commented):
This is an experimental feature that requires additional dependency to run `FaceAnalysis` and it adds a lot of code. If possible, I would like to find a way to support this from the community folder, for this one and similar models moving forward. 

let me know what you guys think! @patrickvonplaten @sayakpaul and @DN6 here

sayakpaul commented:
Hmm, that seems to be the case. However, this could enable precise face editing as well no? I see that to be a real use case and it definitely has demand for it. Could we maybe try to think of a better design (such as delegating to face-specific modules) so that the code doesn't become bloated? 

So, my idea is that we add utilities for face-specific IP Adapters. When loading IP adapters, upon the detection of a face-related adapter, we delegate the rest of the handling to those utilities. 

I don't mind the additional dependency given we do it for ControlNet, too (you cannot run many ControlNet checkpoints with `controlnet_aux`, for example) and precise face editing is a real use case.

yiyixuxu commented:
@sayakpaul, I completely agree that this can enable real use cases and have huge potential.

However, this model was released as an experimental checkpoint, so I think it's too early for us to integrate. I think we have two options here:
1. add this a community pipeline 
2. wait a little bit to see if they release a stable checkpoint and if this model gains more popularity 

I prefer the option 1 - It will require a little bit of effort to move all the files to the community folder but will allow users to try this pipeline out at mealtime . What do you think? @fabiorigano it will be completely up to you:)

fabiorigano commented:
hi @yiyixuxu, no problem for me, I can move everything to the community folder. I would leave some lines in common with IPAdapterFull unchanged, because in my opinion the current PR implementation is more readable and general.

I also think that all classes copying `set_default_attn_processor` from UNet2DConditionModel should be updated to support pytorch's sdpa, as it is done in this PR.

yiyixuxu commented:
@fabiorigano 
 > I would leave some lines in common with IPAdapterFull unchanged, because in my opinion the current PR implementation is more readable and general.
I also think that all classes copying set_default_attn_processor from UNet2DConditionModel should be updated to support pytorch's sdpa, as it is done in this PR.

ok by me - let's review it again once move it to the community folder :)

yiyixuxu commented:
@fabiorigano 
it will be easier if you separate it into two PRs:
* one PR just for the community pipeline, we can merge that quickly
* and the other PR for the other changes you mentioned - it will require more time for us to review

YiYi

fabiorigano commented:
> @fabiorigano 
> it will be easier if you separate it into two PRs:
> * one PR just for the community pipeline, we can merge that quickly
> * and the other PR for the other changes you mentioned - it will require more time for us to review
> 
> YiYi

Ok perfect! Will do it tomorrow

sayakpaul commented:
> However, this model was released as an experimental checkpoint, so I think it's too early for us to integrate

Makes sense, yeah!

okaris commented:
@fabiorigano Thank you for your contribution. I took your branch on a test run but haven't had success loading the FaceID SDXL weights. 

 ip-adapter-faceid_sdxl.bin is being tried to load as a LoRA and throws an error.

ip-adapter-faceid_sdxl_lora.safetensors throws:
`self.ip_pipe.load_ip_adapter("h94/IP-Adapter-FaceID", subfolder=None, weight_name="ip-adapter-faceid_sdxl_lora.safetensors")
File "/root/.pyenv/versions/3.9.18/lib/python3.9/site-packages/huggingface_hub/utils/_validators.py", line 118, in _inner_fn
return fn(*args, **kwargs)
File "/root/.pyenv/versions/3.9.18/lib/python3.9/site-packages/diffusers/loaders/ip_adapter.py", line 164, in load_ip_adapter
unet._load_ip_adapter_weights(state_dict)
File "/root/.pyenv/versions/3.9.18/lib/python3.9/site-packages/diffusers/loaders/unet.py", line 796, in _load_ip_adapter_weights
num_image_text_embeds = state_dict["image_proj"]["latents"].shape[1]
KeyError: 'latents'
Traceback (most recent call last):
  File "/root/.pyenv/versions/3.9.18/lib/python3.9/site-packages/cog/server/runner.py", line 296, in setup
    for event in worker.setup():
  File "/root/.pyenv/versions/3.9.18/lib/python3.9/site-packages/cog/server/worker.py", line 126, in _wait
    raise FatalWorkerException(raise_on_error + ": " + done.error_detail)
cog.server.exceptions.FatalWorkerException: Predictor errored during setup: 'latents'`

fabiorigano commented:
hi @okaris, when diffusers.utils.USE_PEFT_BACKEND is True, this model can't be loaded. To bypass this constant, I made a little change in the definition, so you can put it to False (@yiyixuxu @sayakpaul I would like to know if it is acceptable for you).

I moved everything to community folder

okaris commented:
@fabiorigano do you mind sharing an example how to use that to disable it?

yiyixuxu commented:
hi @fabiorigano 

> when diffusers.utils.USE_PEFT_BACKEND is True, this model can't be loaded.

do you know why it won't work when this constant is True? i.e. which line of code is causing the problem?

yiyixuxu commented on src/diffusers/utils/constants.py:
```diff
@@ -48,7 +48,8 @@
 ) >= version.parse(MIN_TRANSFORMERS_VERSION)
 
-USE_PEFT_BACKEND = _required_peft_version and _required_transformers_version
+_use_peft = _required_peft_version and _required_transformers_version
```
I don't think I understand the change here - this should be equivalent to the code before, no?  cc @sayakpaul

fabiorigano replied:
if you set it to False before creating the pipeline, you don't use the value of `_use_peft` anymore

```
import diffusers
diffusers.utils.USE_PEFT_BACKEND = False
```

fabiorigano commented:
@okaris 

you have the full example in the README

```
import diffusers
diffusers.utils.USE_PEFT_BACKEND = False
```

fabiorigano commented:
@yiyixuxu 

> do you know why it won't work when this constant is True? i.e. which line of code is causing the problem?

here, it is used toch.nn.Linear instead of LoRACompatibleLinear, and so I cannot call set_lora_layer
https://github.com/huggingface/diffusers/blob/5dc34713800306f498f8ecff6158fcd3668032be/src/diffusers/models/attention_processor.py#L180

yiyixuxu commented:
@fabiorigano 
ohh I saw you reverted it - does that mean it's working now? 

cc @sayakpaul here for his expertise 
@fabiorigano needs to set `diffusers.utils.USE_PEFT_BACKEND = False` - what's the best way to go about this? see comments https://github.com/huggingface/diffusers/pull/6276#issuecomment-1887594712

fabiorigano commented:
@yiyixuxu yes, at first I was probably setting it after creating the pipeline, but it works if set before as I did in the README example. Sorry for not checking before committing 😅

yiyixuxu commented:
Alright! great! I think it is good to merge then 
I will wait @sayakpaul to take a final look :)

okaris commented:
Hi @fabiorigano how can I test the community pipeline version before it's merged?

```
huggingface_hub.utils._errors.HfHubHTTPError: 404 Client Error: Not Found for url: https://raw.githubusercontent.com/huggingface/diffusers/main/examples/community/ip_adapter_face_id.py

```

fabiorigano commented:
Hi @okaris, pass the path to the python file as custom_pipeline: 
```
pipeline = DiffusionPipeline.from_pretrained(
    "SG161222/Realistic_Vision_V4.0_noVAE",
    torch_dtype=torch.float16,
    scheduler=noise_scheduler,
    vae=vae,
    custom_pipeline="./path/to/ip_adapter_face_id.py"
)
```

patrickvonplaten commented:
Great job @fabiorigano!

okaris commented:
Is there a way to make this work with the PEFT backend enabled @yiyixuxu @fabiorigano ?
